src/svd/util.py

```diff
@@ -0,0 +1,386 @@
+                    and addr not in periph_modified_filter
+                ):
+                    continue
+                memory[addr] = val
+
+        return memory
+
+    def build_dict(self, selector: BuildSelector = BuildSelector()) -> dict:
+        """Encode device content as a dictionary representation of the registers and content.
+
+        :param selector: selected parts of the device.
+        :returns: content dictionary.
+        """
+        config = {}
+
+        for peripheral in self._device.values():
+            if not selector.is_periph_selected(peripheral):
+                continue
+
+            cfg_periph = {}
+            nonleaf_stack = [cfg_periph]
+            for reg in peripheral.register_iter():
+                reg_depth = len(reg.path)
+                if reg_depth < len(nonleaf_stack):
+                    # We are returning from a nested context.
+                    # Prune any empty supertables on the way up.
+                    for elem in list(nonleaf_stack[reg_depth:]):
+                        for key, val in list(elem.items()):
+                            if not val:
+                                del elem[key]
+
+                    nonleaf_stack = nonleaf_stack[:reg_depth]
+
+                if not reg.leaf:
+                    reg_table = {}
+                    nonleaf_stack[reg_depth - 1][str(reg.path[-1])] = reg_table
+                    nonleaf_stack.append(reg_table)
+                    continue
+
+                assert isinstance(reg, svd.Register)
+
+                if not selector.is_reg_selected(reg):
+                    continue
+
+                if reg.fields:
+                    reg_table = {}
+
+                    for field_name, field in reg.fields.items():
+                        if not selector.is_field_selected(field):
+                            continue
+
+                        try:
+                            reg_table[field_name] = field.content_enum
+                        except LookupError:
+                            # Content does not match any defined enum
+                            reg_table[field_name] = field.content
+
+                    # Include if at least one field was selected
+                    if reg_table:
+                        nonleaf_stack[reg_depth - 1][str(reg.path[-1])] = reg_table
+                else:
+                    # No fields, just a value
+                    nonleaf_stack[reg_depth - 1][str(reg.path[-1])] = reg.content
+
+            # Prune empty tables from the top level config
+            for key, val in list(cfg_periph.items()):
+                if not val:
+                    del cfg_periph[key]
+
+            if cfg_periph:
+                config[peripheral.name] = cfg_periph
+            else:
+                # Ensure a non-empty config file
+                config[peripheral.name] = {}
+
+        return config
+
+    def apply_memory(self, content: dict[int, int]) -> DeviceBuilder:
```

Comment:
Non-blocking: would it make sense to pass a `BuildSelector` here? Without the ability to limit e.g. which peripherals are processed, this function is barely usable with large and/or buggy SVDs.

Reply:
I'm not sure what the most appropriate solution is for this tbh, so let's leave it for a follow up task